src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java

```diff
@@ -11,21 +15,65 @@
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import java.io.IOException;
+import java.text.NumberFormat;
+import java.text.ParseException;
+import java.util.ArrayList;
+import java.util.List;
 
 
 public class ProductDetailsPageServlet extends HttpServlet {
     private ProductDao productDao;
 
+    private CartService cartService;
+
     @Override
     public void init(ServletConfig config) throws ServletException {
         super.init(config);
         productDao = ArrayListProductDao.getInstance();
+        cartService = DefaultCartService.getInstance();
     }
 
     @Override
     protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
-        String productId = request.getPathInfo();
-        request.setAttribute("product", productDao.getProduct(Long.valueOf(productId.substring(1))));
+        Long productId = parseProductId(request);
+        Product product = productDao.getProduct(productId);
+        List<Product> recentlyViewedProducts = (ArrayList<Product>) request.getSession().getAttribute("recentlyViewedProducts");
```

Comment:
This also need to be refactored, but after call. I will show you why the way you did it is not ok